add proper timeouts for the http get calls

Omarabdul3ziz opened the pull request:
### Description
by creating a custom HTTP client and assigning a default value as a timeout based on each operation


### Related Issues
- https://github.com/threefoldtech/zos/issues/2214

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstring

Omarabdul3ziz committed:
add proper timeouts for the http get calls

by creating a custom http client and assign a default value as a timeout
based on each operation

xmonader reviewed (dismissed):
Looks good!

muhamadazmy commented on pkg/environment/config.go:
```diff
@@ -36,12 +41,12 @@ func GetConfig() (base Config, err error) {
 		return
 	}
-	base, err = getConfig(env.RunningMode, baseExtendedURL)
+	base, err = getConfig(env.RunningMode, baseExtendedURL, defaultHttpTimeout)
```
This is good solution of course. But I am wondering if in the future we needed to customize more of the client behaviour that the argument list will keep growing. May be there is a better way to do this 🤷🏼‍♂️

muhamadazmy commented on pkg/flist/flist.go:
```diff
@@ -603,7 +605,12 @@ func (f *flistModule) Unmount(name string) error {
 	md5URL := url + ".md5"
-	resp, err := http.Get(md5URL)
+
+	cl := &http.Client{
```
the http client in Go actually uses http2 which means it can reuse connections. But of course this won't work if we create a client everytime a client is needed.

I suggest that may be we instead share the client, maybe make it an attribute of the flistModule. this way we won't need to recreate everytime.

muhamadazmy commented on pkg/geoip/geoip.go:
```diff
@@ -47,7 +52,11 @@ func getLocation(geoIPService string) (Location, error) {
 	}
 
-	resp, err := http.Get(geoIPService)
+	cl := &http.Client{
```
the geoip pkg is kinda different. Although i don't like global vars may be instead of declaring the timeout as const we can have a global (private) defaultClient in this pkg this way we also don't have to create an instance everytime. and will behave similar to `http.DefaultClient`

muhamadazmy commented on pkg/upgrade/hub/hub.go:
```diff
@@ -86,7 +89,11 @@ func (h *HubClient) Info(repo, name string) (info FList, err error) {
 
 	u.Path = filepath.Join("api", "flist", repo, name, "light")
-	response, err := http.Get(u.String())
+	cl := &http.Client{
```
the HubClient can obviously also use a client attribute instead of re-creating the client. This way it can use the connection reuse features of http2 and saves on memory

Omarabdul3ziz committed:
make the http client more reusable:
- add an httpClient with default timeout as a prop for hub/flist clients
- a global default http client in geoip package
- a new opts struct for the getconfig function

muhamadazmy commented on pkg/environment/config.go:
```diff
@@ -36,12 +45,17 @@ func GetConfig() (base Config, err error) {
 	}
-	base, err = getConfig(env.RunningMode, baseExtendedURL)
+
+	opts := httpClientOption{
```
I was actually trying to push you to pass the Client itself  not config.

muhamadazmy replied:
We don't need an option struct. Just create the client at the top level then pass it around. Imagine now if u added a new flag to the options. Then u will have to go everywhere and modify the client creation to actually use this flag. Right?

But if you create the Client onetime at the top level, then pass it around and later decided to change one of the client options, u can do it only in this one place

muhamadazmy reviewed (changes requested):
- Tests are falling
- Check my comment

Omarabdul3ziz committed:
create the http client in the root function and pass it around

Omarabdul3ziz committed:
fix the nil pointer in tests

muhamadazmy commented on pkg/upgrade/upgrade.go:
```diff
@@ -95,8 +96,10 @@ func Zinit(socket string) func(u *Upgrader) error {
+	hubClient := hub.NewHubClient(defaultHubTimeout)
 	u := &Upgrader{
 		root: root,
+		hub:  *hubClient,
```
when doing a `deref` (like *something) this copies the value. I think it might be better to change the `hub` attribute to `hub *hub.HubClient` instead.

Omarabdul3ziz committed:
change the hub in the upgrader to *hub.HubClient
